src/AutoMapper/Execution/TypeMapPlanBuilder.cs

```diff
@@ -5,548 +5,88 @@ namespace AutoMapper.Execution
-        private readonly ParameterExpression _destination;
-        private readonly ParameterExpression _initialDestination;
-        private readonly TypeMap _typeMap;
-
+        
         public TypeMapPlanBuilder(IConfigurationProvider configurationProvider, TypeMap typeMap)
-        {
-            _configurationProvider = configurationProvider;
-            _typeMap = typeMap;
+        {   
+            ConfigurationProvider = configurationProvider;
+            TypeMap = typeMap;
             Source = Parameter(typeMap.SourceType, "src");
-            _initialDestination = Parameter(typeMap.DestinationTypeToUse, "dest");
+            InitialDestination = Parameter(typeMap.DestinationTypeToUse, "dest");
             Context = Parameter(typeof(ResolutionContext), "ctxt");
-            _destination = Variable(_initialDestination.Type, "typeMapDestination");
+            Destination = Variable(InitialDestination.Type, "typeMapDestination");
         }
 
-        public ParameterExpression Source { get; }
-
-        public ParameterExpression Context { get; }
+        internal TypeMap TypeMap { get; }
+        internal ParameterExpression Source { get; }
+        internal ParameterExpression Destination { get; }
+        internal ParameterExpression InitialDestination { get; }
+        internal IConfigurationProvider ConfigurationProvider { get; }
+        internal ParameterExpression Context { get; }
 
         public LambdaExpression CreateMapperLambda(Stack<TypeMap> typeMapsPath)
         {
-            if (_typeMap.SourceType.IsGenericTypeDefinition() ||
-                _typeMap.DestinationTypeToUse.IsGenericTypeDefinition())
+            if (TypeMap.SourceType.IsGenericTypeDefinition() ||
+                TypeMap.DestinationTypeToUse.IsGenericTypeDefinition())
                 return null;
             var customExpression = TypeConverterMapper() ??
-                                   _typeMap.Substitution ?? _typeMap.CustomMapper ?? _typeMap.CustomProjection;
+                                   TypeMap.Substitution ?? TypeMap.CustomMapper ?? TypeMap.CustomProjection;
             if (customExpression != null)
-                return Lambda(customExpression.ReplaceParameters(Source, _initialDestination, Context), Source,
-                    _initialDestination, Context);
-
-            CheckForCycles(typeMapsPath);
-
-            var destinationFunc = CreateDestinationFunc(out bool constructorMapping);
-
-            var assignmentFunc = CreateAssignmentFunc(destinationFunc, constructorMapping);
-
-            var mapperFunc = CreateMapperFunc(assignmentFunc);
-
-            var checkContext = CheckContext(_typeMap, Context);
+                return Lambda(customExpression.ReplaceParameters(Source, InitialDestination, Context), Source,
+                    InitialDestination, Context);
+
+            this.CheckForCycles(typeMapsPath);
+
+            var mapperFunc = Block(

```

Comment:
This block is now much more readable in what goes on and all the function calls are located in a class that consolidates the functionality.